homeassistant/components/mqtt/mixins.py

```diff
@@ -513,12 +516,165 @@ async def cleanup_device_registry(hass, device_id, config_entry_id):
+
+        # Only activate if the parent class has a discovery hash
+        self.hass = hass
+        self.log_name = log_name
+        self._remove_discovery = None
+
+        self._discovery_data = discovery_data
+        self._device_id = device_id
+        self._config_entry = config_entry
+        self._config_entry_id = config_entry.entry_id
+        self._skip_device_removal: bool = False
+        self._async_update_callback = async_update_callback
+
+        discovery_hash = get_discovery_hash(discovery_data)
+        self._remove_signal = async_dispatcher_connect(
+            hass,
+            MQTT_DISCOVERY_UPDATED.format(discovery_hash),
+            self.async_discovery_update,
+        )
+        config_entry.async_on_unload(self._entry_unload)
+        if device_id is not None:
+            self._remove_device_updated = hass.bus.async_listen(
+                EVENT_DEVICE_REGISTRY_UPDATED, self._async_device_removed
+            )
+        _LOGGER.info(
+            "%s %s has been initialized",
+            self.log_name,
+            discovery_hash,
+        )
+
+    async def async_discovery_update(
+        self,
+        discovery_payload: DiscoveryInfoType | None,
+    ) -> None:
+        """Handle discovery update."""
+        discovery_hash = get_discovery_hash(self._discovery_data)
+        _LOGGER.info(
+            "Got update for %s with hash: %s '%s'",
+            self.log_name,
+            discovery_hash,
+            discovery_payload,
+        )
+        if (
+            discovery_payload
+            and discovery_payload != self._discovery_data[ATTR_DISCOVERY_PAYLOAD]
+        ):
+            _LOGGER.info(
+                "%s %s updating",
+                self.log_name,
+                discovery_hash,
+            )
+            await self.async_update(discovery_payload)
+        if not discovery_payload:
+            # Unregister and clean up the current discovery instance
+            terminate_discovery(self.hass, self._discovery_data, self._remove_signal)
+            await self._async_tear_down()
+            send_discovery_done(self.hass, self._discovery_data)
+            _LOGGER.info(
+                "%s %s has been removed",
+                self.log_name,
+                discovery_hash,
+            )
+        else:
+            # Normal update without change
+            send_discovery_done(self.hass, self._discovery_data)
+            _LOGGER.info(
+                "%s %s no changes",
+                self.log_name,
+                discovery_hash,
+            )
+            return
+
+    async def _async_device_removed(self, event) -> None:
+        """Handle the manual removal of a device."""
+        if self._skip_device_removal or not async_removed_from_device(
+            self.hass, event, self._device_id, self._config_entry_id
+        ):
+            return
+        # Avoid loop of device registry updates
+        self._skip_device_removal = True
```

Comment:
If this is not done, when does the loop of device registry updates happen?

Reply:
Updated the comment to `Prevent a second attempt of the device removal during cleanup`